core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java

```diff
@@ -194,4 +194,170 @@ public void testNoBuffering() throws Exception {
+				"                 sh:value                      ex:steve\n" +
+				"               ] ;\n" +
+				"  sh:result    [ a                             sh:ValidationResult ;\n" +
+				"                 sh:focusNode                  ex:validPerson1 ;\n" +
+				"                 sh:resultPath                 ex:knows ;\n" +
+				"                 sh:resultSeverity             sh:Violation ;\n" +
+				"                 sh:sourceConstraintComponent  sh:NotConstraintComponent ;\n" +
+				"                 sh:sourceShape                _:b0 ;\n" +
+				"                 sh:value                      ex:steve\n" +
+				"               ] ;\n" +
+				"  sh:result    [ a                             sh:ValidationResult ;\n" +
+				"                 sh:focusNode                  ex:validPerson1 ;\n" +
+				"                 sh:resultPath                 ex:knows ;\n" +
+				"                 sh:resultSeverity             sh:Violation ;\n" +
+				"                 sh:sourceConstraintComponent  sh:NotConstraintComponent ;\n" +
+				"                 sh:sourceShape                _:b0 ;\n" +
+				"                 sh:value                      ex:steve\n" +
+				"               ] ;\n" +
+				"  sh:result    [ a                             sh:ValidationResult ;\n" +
+				"                 sh:focusNode                  ex:validPerson1 ;\n" +
+				"                 sh:resultPath                 ex:knows ;\n" +
+				"                 sh:resultSeverity             sh:Violation ;\n" +
+				"                 sh:sourceConstraintComponent  sh:NotConstraintComponent ;\n" +
+				"                 sh:sourceShape                _:b0 ;\n" +
+				"                 sh:value                      ex:steve\n" +
+				"               ]\n" +
+				"] .\n";
+
+		System.out.println("### EXPECTED ###");
+		System.out.println(data);
+		System.out.println("#################\n");
+
+		Model expected = Rio.parse(new StringReader(data), "", RDFFormat.TURTLE);
+
+		StringWriter stringWriter = new StringWriter();
+		WriterConfig config = new WriterConfig();
+		config.set(BasicWriterSettings.INLINE_BLANK_NODES, true);
+		Rio.write(expected, stringWriter, RDFFormat.TURTLE, config);
+
+		System.out.println("### ACTUAL ###");
+		System.out.println(stringWriter.toString());
+		System.out.println("#################\n");
+
+		Model actual = Rio.parse(new StringReader(stringWriter.toString()), "", RDFFormat.TURTLE);
+		assertTrue(Models.isomorphic(expected, actual));
+	}
+
+	@Test
+	public void anotherBnodeTestReduced() throws Exception {
+		String data = "" +
+				"@prefix ex:    <http://example.com/ns#> .\n" +
+				"@prefix sh:    <http://www.w3.org/ns/shacl#> .\n" +
+				"@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n" +
+				"@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .\n" +
+				"@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .\n" +
+				"@prefix foaf:  <http://xmlns.com/foaf/0.1/> .\n" +
+				"\n" +
+				"ex:a  ex:list   (_:b0 _:b0) .";
+
```

Comment:
This seems to be a representative example.

Reply:
This is what RDF4J outputs:

```
@prefix ex: <http://example.com/ns#> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .

ex:a ex:list ([] []) .
```

Reply:
...that is correct, as far as I can tell. In this case, the bug might actually be in the isomorphism detection, because as far as I can tell those two graphs _should_ be isomorphic.

Reply:
Then again, what we're actually after here is no loss of information of course. I'll take a closer look at the writer, see why this node gets inlined when it's referred to more than once.